### ISSUE TITLE
Retry on timeouts?

### ISSUE THREAD

**Issue by kfriedman** [NYPL/hold-request-result-consumer#30]
We had a request to PATCH a HoldRequest (1131) timeout:

https://console.aws.amazon.com/cloudwatch/home?region=us-east-1#logEventViewer:group=/aws/lambda/HoldRequestResultConsumer-production;stream=2017/09/08/%5B$LATEST%5D7c8db76c825846e78cce16d9ec6e687a;refid=33560460446547838035345234291900279714016486518119858184;reftime=1504903093945

Since it wasn't retried, this caused an otherwise successful request to be in limbo and fail in SCSB. 

Should we consider making timeouts a retryable exception?

**Comment by holingpoon** [comment]
@kfriedman The time out message below. There is no particular status code, i.e. the error code is 0, so the approach may be to do a `strpos()` on `message` within the `catch (Exception e)` block. Are we only retrying specific time outs?
```2017-09-08T20:38:13.944Z	3acfa1fb-9a53-4a61-9ee0-8525db468d61
{
    "message": "Exception thrown: cURL error 28: Operation timed out after 10000 milliseconds with 0 bytes received (see http://curl.haxx.se/libcurl/c/libcurl-errors.html), Error code: 0",
    "level": "ERROR",
    "datetime": {
        "date": "2017-09-08 16:38:13.877629",
        "timezone_type": 3,
        "timezone": "America/New_York"
    },
    "levelCode": 3,
    "timestamp": "2017-09-08T16:38:13-04:00"
}

**Comment by kfriedman** [comment]
Hmmm... do you have any idea what exception Guzzle is throwing for a timeout?

In any case, I doubt it's a specific timeout exception so I think your approach (i.e. some kind of string matching on the `message`) sounds good. I would retry on all timeouts unless you think that might pose some kind of problem.

**Comment by holingpoon** [comment]
I did a trace and found out that it's a `GuzzleHttp\Promise\RejectedPromise` object created and returned by this method:
`GuzzleHttp\Handler\CurlFactory::createRejection(EasyHandler,array)`
No signs of an error code is being returned. I think we should stick with `strpos` approach. I'm okay with all the other `Operation timed out` messages getting free retries anyway. 🙂

**Comment by holingpoon** [comment]
Merged PR #31 so closing this issue 😄
